controllers/userControllers.js

```diff
@@ -1,78 +1,147 @@
+      }
+    } catch (error) {
+      next(error);
     }
+  }
 
-    async updateUser(req, res, next) {
-
-        const updates = Object.keys(req.body) 
-        const allowedUpdates = ['firstName', 'lastName', 'email', 'photo', 'password', 'roleId']
-        const isValidOperation = updates.every((update) => allowedUpdates.includes(update))
-        
-        try {
-            const adminQuery = await User.findAll({
-                where: {
-                    id: req.params.id
-                }
-            })
-            
-            if(!isValidOperation || !adminQuery) return res.status(400).send({ error: 'Invalid update!' })
-            else if(isValidOperation ) {
-              
-                const query = await User.update( req.body, { 
-                    where: {
-                        id: parseInt(req.params.id)
-                }})
-    
-                res.json(query)
-            }
-        } catch (error) {
-            next(error)
-        }
+  async deleteUser(req, res, next) {
+    try {
+      const adminQuery = await User.findAll({
+        where: {
+          id: req.params.id,
+        },
+      });
+
+      if (adminQuery) {
+        const data = await User.destroy({
+          where: {
+            id: req.params.id,
+          },
+        });
+        if (data)
+          return res.json({ msge: 'The user has been successfully deleted' });
+        else
+          return res
+            .status(404)
+            .json({ msge: 'An error has occured. The user doesnt exist' });
+      }
+    } catch (error) {
+      next(error);
     }
+  }
+
+  async patchUser(req, res, next) {
+    var user, result;
+
+    const errors = validationResult(req);
+
+    if (!errors.isEmpty()) {
+      return res
+        .status(HttpStatusCodes.BAD_REQUEST)
+        .json({ errors: errors.array() });
+    } else {
+      let userId = req.params.id;
 
-    async deleteUser(req, res, next) {
-        try {
-            const adminQuery = await User.findAll({
-                where: {
-                    id: req.params.id
-                }
-            })
-    
-            if(adminQuery) {
-                const data = await User.destroy({
-                    where: {
-                        id: req.params.id
-                    }
-                })
-                if(data) return res.json({msge: 'The user has been successfully deleted'})
-                else return res.status(404).json({msge: 'An error has occured. The user doesnt exist'})
-                
-            }
-            
-        } catch (error) {
-            next(error)
+      try {
+        user = await User.findByPk(userId);
+
+        if (!user) {
+          throw new Error('User not found');
         }
+      } catch (error) {
+        return res
+          .status(HttpStatusCodes.NOT_FOUND)
+          .send({ status: error.message });
+      }
+
+      let firstName = req.body.firstName ? req.body.firstName : user.firstname;
```

Comment:
nit: Para hacer más limpio esto lo llevaria a una función aparte.

Reply:
¿Algo así como esto? ¿O la sacarías a la clase controlador? ¿O a otra clase?

  async patchUser(req, res, next) {
    var user, result, firstName, lastName, email, image, password, roleId;

    function assignValues() {
      firstName = req.body.firstName ? req.body.firstName : user.firstname;
      lastName = req.body.lastName ? req.body.lastName : user.lastName;
      email = req.body.email ? req.body.email : user.email;
      image = req.body.image ? req.body.image : user.image;
      password = req.body.password ? req.body.password : user.password;
      roleId = req.body.roleId ? req.body.roleId : user.roleId;
    }

    const errors = validationResult(req);

    if (!errors.isEmpty()) {
      return res
        .status(HttpStatusCodes.BAD_REQUEST)
        .json({ errors: errors.array() });
    } else {
      let userId = req.params.id;

      try {
        user = await User.findByPk(userId);

        if (!user) {
          throw new Error('User not found');
        }
      } catch (error) {
        return res
          .status(HttpStatusCodes.NOT_FOUND)
          .send({ status: error.message });
      }

      assignValues();

      try {
        result = await User.update(
          {
            firstName: firstName,
            lastName: lastName,
            email: email,
            image: image,
            password: password,
            roleId: roleId,
          },
          {
            where: { id: userId },
          }
        );
      } catch (error) {
        return res
          .status(HttpStatusCodes.INTERNAL_SERVER_ERROR)
          .send({ status: error.message });
      }

      return res.send({ 'Updated records: ': result });
    }
  }

Reply:
No, dentro del scope de la clase para que sea reutilizable no dentro de la función(no es reusable).